PEP 819: JSON Package Metadata (#4751)
python/peps: peps/pep-0819.rst

PEP: 819
Title: JSON Package Metadata
Author: Emma Harper Smith <emma@python.org>
PEP-Delegate: Paul Moore
Discussions-To: Pending
Status: Draft
Type: Standards Track
Topic: Packaging
Created: 18-Dec-2025
Post-History: Pending

Abstract
========

Python package metadata ("core metadata") was first defined in :pep:`241` to
use :rfc:`822` email headers to encode information about packages. This was
reasonable at the time; email messages were the only widely used, standardized
text format that had a parser in the standard library at the time. However,
issues with handling different encodings, differing handling of line breaks,
and other differences between implementations have caused numerous packaging
bugs. To resolve these issues, this PEP proposes introducing
`Javascript Object Notation (JSON) <https://www.json.org/json-en.html>`_

> encoded core metadata and wheel file format metadata files in Python packages.

Reviewer: A suggested rewording to lead with the actual change that's being proposed. I think the text justifying the original choice of email headers could be moved to the body of the PEP (& removed from the Abstract), but that's an aesthetic choice.

We can and should assume at least a 'generically technical' reader for PEPs, so e.g. expanding & linking to JSON's website is unnecessary.

```suggestion
This PEP proposes introducing JSON encoded core metadata and wheel file format
metadata files in Python packages. Python package metadata ("core metadata")
was first defined in :pep:`241` to use :rfc:`822` email headers to encode
information about packages. This was reasonable in 2001; email messages
were the only widely used, standardized text format that had a parser in
the standard library. However, issues with handling different encodings,
differing handling of line breaks, and other differences between
implementations have caused numerous packaging bugs. Using the JSON format for
encoding metadata files would eliminate a wide range of these potential issues.
```

Motivation
==========

The email message format has a number of complexities and limitations which
reduce its utility as a portable textual interchange format for packaging
metadata. Due to the :mod:`email` parser requiring configuration changes to
properly generate valid core metadata, many projects do not use the
:mod:`!email` module and instead generate core metadata in a custom manner.
There are many pitfalls with generating email headers that these custom
generators can hit. First, core metadata fields may contain newlines in the
value of fields. These newlines must be handled properly to "unfolded" multiple
lines per :rfc:`822`. One particularly difficult to encode field is the
``Description`` field, which may contain newlines and indentation. To encode
the field in email headers, CRLF line breaks must be followed by seven (7)

> spaces and a pip ("|") character. While ``Description`` may now be encoded in

Reviewer: The 'official' Unicode name is VERTICAL LINE, could use that in preference to 'pipe'
```suggestion
spaces and a pipe ('``|``') character. While ``Description`` may now be encoded in
```

Author: I believe I used pipe here to mirror the text on packaging.python.org, so I'll probably keep pipe to be consistent.

the message body, similar escaping issues occur for the ``Author`` and
``Maintainer`` fields. Improperly escaped newlines can lead to missing,
partial, or invalid core metadata. Second, as discussed in the
`core metadata specifications <https://packaging.python.org/specifications/core-metadata/>`__:

> .. epigraph::

Reviewer: Per [docs](https://docutils.sourceforge.io/docs/ref/rst/directives.html#epigraph), this has no argument, so should be followed by a blank line
```suggestion
.. epigraph::

```

    The standard file format for metadata (including in wheels and installed
    projects) is based on the format of email headers. However, email formats
    have been revised several times, and exactly which email RFC applies to
    packaging metadata is not specified. In the absence of a precise
    definition, the practical standard is set by what the standard library
    :mod:`email.parser` module can parse using the
    :data:`email.policy.compat32` policy.

Since no specific email RFC is selected, the current core metadata
specification is ambiguous whether a given core metadata document is valid.
:rfc:`822` is the only email standard to be explicitly listed in a PEP.
However, the core metadata specifications also requires that core metadata is
encoded using UTF-8 when written to a file. This de-facto makes the core
metadata follow :rfc:`6532`, which specifies internationalization of email

> headers. This has practical interoperability concerns. Until a few years ago,

Reviewer: Could we be more precicse here?

Author: Oh perhaps you mean "a few years ago". In which case I believe this would be the proper change.
```suggestion
headers. This has practical interoperability concerns. Until three years ago,
```
(ref https://github.com/pypa/packaging.python.org/pull/1046)

it was unspecified how to handle non-ASCII encoded content in core metadata,
causing confusion about how to properly encode non-ASCII emails in core

> metadata. Third, the current format is difficult to properly validate and

Reviewer: This seems repetitive: unspecified how to handle non-ASCII causing confusion about how to handle non-ASCII.

parse. Many tools do not check for issues with the output of the :mod:`!email`
parser. If a document is malformed, it may still parse without error by the
:mod:`!email` module as a valid email message. Furthermore, due to limitations
in the email format, fields like ``Project-Url`` must create custom encodings
of nested key-value items, further complicating parsing.  Finally, the lack of
a schema makes it difficult to validate the contents of email message encoded
metadata. While introducing a specification for the current format has been

> `discussed previously <https://discuss.python.org/t/python-metadata-format-specification-and-implementation/7550>`_,

Reviewer: ```suggestion
`discussed previously <https://discuss.python.org/t/7550>`__,
```

performance, this equivalence is not required to be verified by installers,
though other tools may do so. Some tools may choose to make the check dependent
on a configuration flag.

Package indexes SHOULD check that the metadata files are semantically
equivalent when the package is added to the index. This is a low-cost, one-time
check that ensures users of the index are served valid packages.

Specification
=============

JSON Format Core Metadata File
------------------------------

A new optional but recommended file ``METADATA.json`` shall be introduced as a
metadata file for Python packages. If generated, the ``METADATA.json`` file
MUST be placed in the same directory as the current email formatted
``METADATA`` or ``PKG-INFO`` file.

For wheels, this means that ``METADATA.json`` MUST be located in the
``.dist-info`` directory. The wheel format minor version will be incremented to
indicate the change in the format.

For source distribution packages, the ``METADATA.json`` file MUST be located
in the root directory of the project sources. Tools that prefer the JSON
formatted metadata file MUST check for the existence of a ``METADATA.json``

> in the source distribution before reading the file.

Reviewer: This seems circular?
